### DISCUSSION TITLE
How to auto-run an external program after Pytest completes execution?

### DISCUSSION THREAD

**Discussion by jeffwright13** [pytest-dev/pytest#9595] [Q&A] [answered]
I've written [a Pytest plugin](https://github.com/jeffwright13/pytest-fold) that marks up sections of the output for use by an external program after the test run is over. How can I get Pytest to automatically invoke the external program after it is done?

For this use case, I want to run Pytest with my plugin's `fold` option, and then have it launch my [TUI code](https://github.com/jeffwright13/pytest-fold/blob/main/pytest_fold/tui.py) (which is written in Python) after it has completed its test run. Any ideas?

**Comment by nicoddemus** [comment]
Hi @jeffwright13, 

Perhaps `pytest_sesssionfinish` is what you are looking for?

**Reply by jeffwright13** [reply]
I'll look into it

**Reply by jeffwright13** [reply]
I added the following code to my `plugin.py`:
```
@pytest.hookimpl(trylast=True)
def pytest_sessionfinish(session: Session):
    subprocess.run(
        ["python", "/Users/jwr003/coding/pytest-fold/pytest_fold/tui.py"]
    )
```
The program seems to launch, but unfortunately it causes the terminal to corrupt and there is a bunch of error code spit to the screen:
```
Traceback (most recent call last):
File "/Users/jwr003/coding/pytest-fold/pytest_fold/tui.py", line 187, in <module> main()
File "/Users/jwr003/coding/pytest-fold/pytest_fold/tui.py", line 180, in main Screen.wrapper(demo, catch_interrupt=True, arguments=[last_scene])
File "/Users/jwr003/coding/pytest-fold/venv/lib/python3.9/site-packages/asciimatics/screen.py", line 1386, in wrapper screen = Screen.open(height,
File "/Users/jwr003/coding/pytest-fold/venv/lib/python3.9/site-packages/asciimatics/screen.py", line 1355, in open screen = _CursesScreen(stdscr, height,
File "/Users/jwr003/coding/pytest-fold/venv/lib/python3.9/site-packages/asciimatics/screen.py", line 2354, in __init__ tbs = termios.tcgetattr(sys.stdin)[6][termios.VERASE]
termios.error: (19, 'Operation not supported by device') %
```

It's also clear that despite the claim in the docs that pytest_sessionfinish is "Called after whole test run finished, right before returning the exit status to the system" - there is a lot of other code called after that, because when I set a break point and follow it into a stub subprocess, it returns and then runs dozens of lines of other code. 

Anyway, I'll keep digging.

**Reply by nicoddemus** [reply]
> It's also clear that despite the claim in the docs that pytest_sessionfinish is "Called after whole test run finished, right before returning the exit status to the system" - there is a lot of other code called after that

That's expected: it is a hook, so it will invoke all other `pytest_sessionfinish` hooks in the system (internals and plugins). Those hooks can run anything, plus there's the code involved in the hook machinery itself (`pluggy`).

**Reply by jeffwright13** [reply]
Cool, thanks - I figured it was something like that. Looks like I have some learning to do about terminal handling...

**Comment by jeffwright13** [comment, marked as answer]
I ended up just calling my own method at the end of the whole test run (end of `pytest_unconfigure`), which makes a subprocess call to my TUI code. For now I seem to have solved the problem. Thanks for your input everyone. :-)
